internal/compiler/checkerpool.go

```diff
@@ -0,0 +1,90 @@
+package compiler
+
+import (
+	"context"
+	"iter"
+	"slices"
+	"sync"
+
+	"github.com/microsoft/typescript-go/internal/ast"
+	"github.com/microsoft/typescript-go/internal/checker"
+	"github.com/microsoft/typescript-go/internal/core"
+)
+
+type CheckerPool interface {
+	GetChecker(ctx context.Context) (*checker.Checker, func())
+	GetCheckerForFile(ctx context.Context, file *ast.SourceFile) (*checker.Checker, func())
+	GetAllCheckers(ctx context.Context) ([]*checker.Checker, func())
+	Files(checker *checker.Checker) iter.Seq[*ast.SourceFile]
+}
+
+type checkerPool struct {
+	checkerCount int
+	program      *Program
+
+	createCheckersOnce sync.Once
+	checkers           []*checker.Checker
+	fileAssociations   map[*ast.SourceFile]*checker.Checker
+}
+
+var _ CheckerPool = (*checkerPool)(nil)
+
+func newCheckerPool(checkerCount int, program *Program) *checkerPool {
+	pool := &checkerPool{
+		program:      program,
+		checkerCount: checkerCount,
+		checkers:     make([]*checker.Checker, checkerCount),
+	}
+
+	return pool
+}
+
+func (p *checkerPool) GetCheckerForFile(ctx context.Context, file *ast.SourceFile) (*checker.Checker, func()) {
+	p.createCheckers()
+	checker := p.fileAssociations[file]
+	return checker, noop
+}
+
+func (p *checkerPool) GetChecker(ctx context.Context) (*checker.Checker, func()) {
+	p.createCheckers()
+	checker := p.checkers[0]
+	return checker, noop
+}
+
+func (p *checkerPool) createCheckers() {
+	p.createCheckersOnce.Do(func() {
+		wg := core.NewWorkGroup(p.program.singleThreaded())
+		for i := range p.checkerCount {
+			wg.Queue(func() {
+				p.checkers[i] = checker.NewChecker(p.program)
+			})
+		}
+
+		wg.RunAndWait()
+
+		p.fileAssociations = make(map[*ast.SourceFile]*checker.Checker, len(p.program.files))
```

Comment:
this is good for now but in theory the questions are asked only on open files so assign only for those ?

Reply:
I meant in editor scenario not cli

Reply:
This implementation is only used in the CLI. In the editor this map is created with no capacity hint.